File: ruby/create_domain.rb
#!/usr/bin/env ruby

# Usage Sandbox: create_domain.rb example.com
# Usage Prod: create_domain.rb example.com

require 'pp'
require 'dnsimple'
require_relative 'token'

base_url =  'https://api.sandbox.dnsimple.com'
ARGV.each do |arg|
  base_url = nil if arg.downcase == 'prod' || arg.downcase == 'production'
end

# Construct a client instance.
#
# If you want to connect to production, add command argument `prod` before the domain argument.
client = Dnsimple::Client.new(base_url: base_url, access_token: TOKEN)

# All calls to client pass through a service. In this case, `client.identity` is the identity service.
#
# Dnsimple::Client::Identity#whoami is the method for retrieving the account details for your
# current credentials via the DNSimple API.
response = client.identity.whoami

# The response object returned by any API method includes a `data` attribute. Underneath that
# attribute is an attribute for each data object returned. In this case, `#account` provides
# access to the resulting account object.
#
# Here the account ID is extracted for use in the call to create domain.
account_id = response.data.account.id

name = ARGV.last
puts "Creating domain #{name}"

# Dnsimple::Client::Domains#create_domain is the method for a new domain in DNSimple. Note that this
# does not register the domain, it simply adds the domain for management and DNS service.
response = client.domains.create_domain(account_id, name: name)

# Pretty-print the entire response object so you can see what is inside.
pp response

File: golang/cmd/create_domain/create_domain.go
// Usage: TOKEN=token go run cmd/create_domain/create_domain.go example.com
package main

import (
	"context"
	"fmt"
	"os"
	"strconv"

	"github.com/dnsimple/dnsimple-go/dnsimple"
	"golang.org/x/oauth2"
)

func main() {
	oauthToken := os.Getenv("TOKEN")
	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: oauthToken})
	tc := oauth2.NewClient(context.Background(), ts)

	client := dnsimple.NewClient(tc)
	client.BaseURL = "https://api.sandbox.dnsimple.com"

	// get the current authenticated account (if you don't know who you are)
	whoamiResponse, err := client.Identity.Whoami()
	if err != nil {
		fmt.Printf("Whoami() returned error: %v\n", err)
		os.Exit(1)
	}

	accountId := strconv.Itoa(whoamiResponse.Data.Account.ID)

	name := os.Args[1]
	fmt.Printf("Adding domain %v\n", name)
	domain := dnsimple.Domain{Name: name}
	createDomainResponse, err := client.Domains.CreateDomain(accountId, domain)
	if err != nil {
		fmt.Printf("CreateDomain() returned error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("%+v\n", createDomainResponse.Data)
}
